### PR TITLE
expression, plan: rewrite builtin func found_rows, database, current_user, user, connection_id, version

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL @zz-jason @breeswish

### CODE REVIEW

File: expression/builtin_info.go [hunk 1]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
-func (b *builtinDatabaseSig) eval(_ []types.Datum) (d types.Datum, err error) {
+func (b *builtinDatabaseSig) evalString(row []types.Datum) (string, bool, error) {
 	currentDB := b.ctx.GetSessionVars().CurrentDB
 	if currentDB == "" {
-		return d, nil
+		return "", true, nil
 	}
-	d.SetString(currentDB)
-	return d, nil
+	return currentDB, false, nil
```
Comment (zz-jason) [inline expression/builtin_info.go]: how about `return currentDB, currentDB == "", nil` and remove the above `if` statement

**Comment by zz-jason** [comment]
@XuHuaiyu plz resolve conflicts.
@breeswish PTAL.

File: expression/builtin_info.go [hunk 2]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
 }
 
 type builtinFoundRowsSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinFoundRowsSig.
+// evalInt evals a builtinFoundRowsSig.
 // See https://dev.mysql.com/doc/refman/5.6/en/information-functions.html#function_found-rows
```
Comment (winkyao) [inline expression/builtin_info.go]: s/5.6/5.7/g ?

File: expression/builtin_info.go [hunk 3]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
 type foundRowsFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *foundRowsFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinFoundRowsSig{newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err := errors.Trace(c.verifyArgs(args)); err != nil {
```
Comment (winkyao) [inline expression/builtin_info.go]: merge master and modify `getFunction(args []Expression, ctx context.Context)` --> `getFunction(ctx context.Context, args []Expression)`

File: expression/builtin_info.go [hunk 4]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
 type currentUserFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *currentUserFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinCurrentUserSig{newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err := errors.Trace(c.verifyArgs(args)); err != nil {
```
Comment (winkyao) [inline expression/builtin_info.go]: ditto.

File: expression/builtin_info.go [hunk 5]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
 type userFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *userFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinUserSig{newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err := errors.Trace(c.verifyArgs(args)); err != nil {
```
Comment (winkyao) [inline expression/builtin_info.go]: ditto.

File: expression/builtin_info.go [hunk 6]
```diff
@@ -58,133 +58,146 @@ type databaseFunctionClass struct {
...
 type connectionIDFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *connectionIDFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinConnectionIDSig{newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err := errors.Trace(c.verifyArgs(args)); err != nil {
```
Comment (winkyao) [inline expression/builtin_info.go]: ditto.

**Comment by XuHuaiyu** [comment]
PTAL @winkyao @zz-jason

**Comment by zz-jason** [comment]
LGTM

**Comment by winkyao** [comment]
please fix ci. @XuHuaiyu

**Pull request merged**